clients/application-kafka/src/main/java/org/eclipse/hono/application/client/kafka/impl/KafkaApplicationClientFactoryImpl.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) 2021 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+package org.eclipse.hono.application.client.kafka.impl;
+
+import java.util.Objects;
+import java.util.function.Consumer;
+
+import org.eclipse.hono.application.client.DownstreamMessage;
+import org.eclipse.hono.application.client.MessageConsumer;
+import org.eclipse.hono.application.client.kafka.KafkaApplicationClientFactory;
+import org.eclipse.hono.kafka.client.consumer.KafkaConsumerConfigProperties;
+
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.kafka.client.consumer.KafkaConsumer;
+
+/**
+ * A factory for creating clients for Hono's Kafka-based northbound APIs.
+ */
+public class KafkaApplicationClientFactoryImpl implements KafkaApplicationClientFactory {
+
+    private final Vertx vertx;
+    private final KafkaConsumerConfigProperties config;
+
+    /**
+     * Creates a new factory.
+     *
+     * @param vertx The Vert.x instance to use.
+     * @param config The Kafka consumer configuration properties to use.
+     * @throws NullPointerException if any of the parameters is {@code null}.
+     * @throws IllegalArgumentException if config does not contain Kafka configuration properties.
+     */
+    public KafkaApplicationClientFactoryImpl(final Vertx vertx, final KafkaConsumerConfigProperties config) {
+        Objects.requireNonNull(vertx);
+        Objects.requireNonNull(config);
+
+        if (!config.isConfigured()) {
+            throw new IllegalArgumentException("No Kafka configuration found!");
+        }
+
+        this.vertx = vertx;
+        this.config = config;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @return a future indicating the outcome. When {@link TelemetryConsumer#start()} completes successfully, the
+     *         future will be completed with the consumer. Otherwise the future will fail with the cause.
+     */
+    @Override
+    public Future<MessageConsumer<DownstreamMessage>> createTelemetryConsumer(final String tenantId,
+            final Consumer<DownstreamMessage> telemetryConsumer, final Handler<Throwable> closeHandler) {
```

Comment:
it looks strange to me that we use both `java.util.function.Consumer` and `io.vertx.core.Handler` in a single method. They serve the same purpose so IMHO we should either use Consumer or Handler but not both in parallel.
I also do not think that the closeHandler accepting a Throwabe provides much benefit because (a) the client has no clue what exceptions to expect and should be able to handle and (b) the whole point of introducing the abstract consumers is to decouple the client from the messaging infrastructure implementation in the first place, isn't it?
FMPOV we should replace it with a `Consumer<Void` or `Handler<Void>` and thus allow the client to react to the channel having been closed in a generic way.

Reply:
> it looks strange to me that we use both `java.util.function.Consumer` and `io.vertx.core.Handler` in a single method. They serve the same purpose so IMHO we should either use Consumer or Handler but not both in parallel.

You are right.

> I also do not think that the closeHandler accepting a Throwabe provides much benefit because (a) the client has no clue what exceptions to expect and should be able to handle 

I'm not sure that this will work well. In the Kafka consumers, there are a couple of different error cases to consider which cannot necessarily be handled in the same way. I tried to group them here: https://github.com/eclipse/hono/blob/7a9e374af0a2f1acd81aa433c951efab0d9fc657/clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/consumer/AbstractAtLeastOnceKafkaConsumer.java#L55
Starting a new consumer might help in some cases (e.g. if a consumer group rebalance takes very long, the offset commit may fail but succeed later on) but not in others (e.g. if the message handler throws an exception, it will probably be thrown on each consumer restart). Therefore I don't know what a completely generic error handling would look like. 
Maybe we should focus on informing the client on how to handle the errors. 

> and (b) the whole point of introducing the abstract consumers is to decouple the client from the messaging infrastructure implementation in the first place, isn't it?

That's what I aim for, but I think there are cases where it might be useful to have more control than the lowest common denominator offers.

Reply:
But is there anything that could be done by client code to remedy the situation, given that it has very limited access to the underlying kafka-client anyway? If I stick to the API defined so far and register a close handler that can accept a Throwable, what can I actually do about it?

Reply:
I don't have the silver bullet here but I think that because the cause of the error is passed to the client application, it can do as much or as little as if it were directly controlling the underlying Kafka consumer. Thinkable options are retrying (by opening a new consumer), stopping the application, triggering an alarm, letting the Kubernetes health check fail, putting the message to a dead-letter-queue. Besides that, it could decide to write the error to logs, traces, or metrics.
FMPOV the critical point is to decide whether retrying or skipping a message makes sense which depends on the type of error. Retries do not come for free. If I have a consumer that polls 100 messages from a topic and fails to process the last one or, similarly, to commit the offset afterwards, then all the messages from that batch will be processed again with each retry. So, with simply restarting, the client application would constantly get the same 100 messages in a loop. Alternatively, if the using application does not know the cause, it could only completely stop processing messages, while there are a lot of potentially temporary root causes.

Some common error handling strategies for the handling of messages consumed from Kafka are provided by [Kafka Connect](https://www.confluent.io/blog/kafka-connect-deep-dive-error-handling-dead-letter-queues/) and, similarly, [Quarkus](https://quarkus.io/blog/kafka-failure-strategy/). They are:

1. Fail fast
2. Ignore message
3. Dead letter queue

I could imagine that at a later stage, when we have more experience with Kafka-based messaging in Hono and when the need becomes apparent, we could offer such strategies directly in the client. Until then, I would take hints in the documentation and leave the implementation to the using application.

Reply:
@sophokles73 I added documentation to the interface `KafkaApplicationClientFactory` about the error cases to expect in the close handler. Does this improve the situation from your POV?